### PR TITLE
Track use of SIMD and exceptions in IR, and CWriter includes support only if used

### PR DESCRIPTION
**Pull request by keithw**
This is tasks 4 and 5 from #2224.

### CODE REVIEW

**Comment by keithw** [comment]
Another way to do this (which would be more systematic but a bigger change) would be to go through the BinaryReader and replace queries to `options.features_.[something]_enabled()` with a call to a new function that checks if the feature is enabled and, if so, does a callback to the BinaryReaderDelegate (e.g. the BinaryReaderIR) tellling it that that feature is in use. Happy to do it this way if anybody thinks better, but I think the way here is probably less intrusive.

File: src/c-writer.cc [hunk 1]
```diff
@@ -1791,10 +1791,10 @@ void CWriter::WriteCallIndirectFuncDeclaration(const FuncDeclaration& decl,
 }
 
 void CWriter::WriteFeatureMacros() {
-  if (options_.features->exceptions_enabled()) {
+  if (module_->features.exceptions) {
```
Comment (shravanrn) [inline src/c-writer.cc:1794]: I thought the plan was to check both the command line switch and if the feature was in use?

File: src/binary-reader-ir.cc [hunk 2]
```diff
@@ -526,6 +526,15 @@ Result BinaryReaderIR::OnFuncType(Index index,
   auto func_type = std::make_unique<FuncType>();
   func_type->sig.param_types.assign(param_types, param_types + param_count);
   func_type->sig.result_types.assign(result_types, result_types + result_count);
+
+  module_->features.simd |=
```
Comment (shravanrn) [inline src/binary-reader-ir.cc:530]: I'm don't understand the code/standard well enough to know if the places this condition is tracked is sufficient. The upside I think is that even if we miss one, that is not security bug and easy enough to fix. So this works. Does FuncTypes also cover externrefs etc.?

**Review by shravanrn — CHANGES_REQUESTED** [review]

File: src/c-writer.cc (hunk 1 above)
Comment (keithw) [reply src/c-writer.cc:1794]: The command-line switches are enforced at read-time, so this can't get set unless it's enabled...

Comment (shravanrn) [reply src/c-writer.cc:1794]: ah gotcha

**Review by shravanrn — APPROVED** [review]

File: include/wabt/ir.h [hunk 3]
```diff
@@ -1245,6 +1245,12 @@ struct Module {
   BindingHash memory_bindings;
   BindingHash data_segment_bindings;
   BindingHash elem_segment_bindings;
+
+  // Feature tracking for a subset of supported features
```
Comment (sbc100) [inline include/wabt/ir.h:1249]: Perhaps be a little more explicit about why this separate from supported features.   How about "For a subset of features we track whether or not they are actually used in the module.  This is used by wasm2c to limit the output in certain cases."

And maybe call it `features_used` or something like that?

**Review by sbc100 — APPROVED** [review]

File: include/wabt/ir.h (hunk 3 above)
Comment (keithw) [reply include/wabt/ir.h:1249]: Sure, done.

**Pull request merged**